atlasdb-autobatch/src/main/java/com/palantir/atlasdb/autobatch/DisruptorAutobatcher.java

```diff
@@ -98,19 +106,60 @@ public void close() {
 
     private static final class DefaultBatchElement<T, R> implements BatchElement<T, R> {
         private T argument;
-        private SettableFuture<R> result;
+        private DisruptorFuture<R> result;
 
         @Override
         public T argument() {
             return argument;
         }
 
         @Override
-        public SettableFuture<R> result() {
+        public DisruptorFuture<R> result() {
             return result;
         }
     }
 
+    public static final class DisruptorFuture<R> extends AbstractFuture<R> {
+
+        private final DetachedSpan parent;
+        private final DetachedSpan waitingSpan;
+
+        @Nullable
+        private DetachedSpan runningSpan = null;
```

Comment:
Does this need to be volatile? Specifically I'm thinking of a race condition where one thread assigns `runningSpan`, but another cancels the future and runs the listener: thread 2 won't see thread 1's write. Maybe this is ok if GC is sufficient to clean out 'leaked' spans

Reply:
nah, this is only updated in the disruptor thread, so it's safe, see the `TracingEventHandler`.